rust/src/shards/gui/containers/docking.rs

```diff
@@ -0,0 +1,460 @@
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      self.contents.compose(&data)?;
+    }
+
+    // Always passthrough the input
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+    self.title.warmup(ctx);
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+    self.title.cleanup();
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      // no contents, same as inactive
+      return Ok(false.into());
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)?;
+
+      // Always passthrough the input
+      Ok(*input)
+    } else {
+      Err("No UI parent")
+    }
+  }
+}
+
+impl Default for DockArea {
+  fn default() -> Self {
+    let mut ctx = ParamVar::default();
+    ctx.set_name(CONTEXTS_NAME);
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      instance: ctx,
+      requiring: Vec::new(),
+      contents: ParamVar::default(),
+      parents,
+      exposing: Vec::new(),
+      headers: Vec::new(),
+      shards: Vec::new(),
+      tabs: egui_dock::Tree::new(Vec::new()),
+    }
+  }
+}
+
+impl Shard for DockArea {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.DockArea")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.DockArea-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.DockArea"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("TODO."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&DOCKAREA_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => {
+        let seq = Seq::try_from(value)?;
+        for tab_spec in seq.iter().rev().filter_map(|s| {
+          if let Ok(r) = ShardRef::try_from(s) {
+            if r.name() == TAB_NAME {
+              let mut title = ParamVar::default();
+              title.set_param(&r.get_parameter(0));
+              let mut shard = ShardsVar::default();
+              match shard.set_param(&s) {
+                Ok(_) => Some(Ok((title, shard))),
```

Comment:
Call it `(title, contents)` like below. Maybe rename the struct member too: shards => contents

Reply:
There is already a struct member named contents (which is the parameter).